### PR TITLE
:sparkles: Feat(Diag): Save the final msg to aigverse

### PR DESCRIPTION
**Pull request by wjrforcyber**
# What
So the messages are collected here for aigverse.

# Why
Related to issue marcelwa/aigverse#130

# How
Since all final messages are printed and discarded after a single call to `report` function (trigger in `handle_diagnostic` function at the end), the messages of different level are not collected, so every time `emit` function is called, we collect the id, args and finally form all the messages.

### CODE REVIEW

**Comment by wjrforcyber** [comment]
Hi @marcelwa , I can't assign reviewer, comment to let you know this PR is ready.

File: lib/lorina/lorina/diagnostics.hpp [hunk 1]
```diff
@@ -254,6 +273,12 @@ inline uint64_t diagnostic_engine::get_num_diagnostics() const
   return num_diagnostics;
 }
 
+inline std::string diagnostic_engine::get_all_messages()
+{
```
[the remark this answers is not in the document]
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp:277]: Done.

File: lib/lorina/lorina/diagnostics.hpp [hunk 2]
```diff
@@ -320,4 +345,28 @@ inline void diagnostic_engine::emit( diag_id id, std::vector<std::string> const&
...
 
+inline void diagnostic_engine::collect_single_msg( diag_id id, std::vector<std::string> const& args )
+{
+    std::pair<diag_id, std::vector<std::string>> collects_ids_args;
+    collects_ids_args.first = id;
+    collects_ids_args.second = args;
+    original_diag_ids_args.push_back( collects_ids_args );
+}
+
+inline void diagnostic_engine::collect_messages( std::vector<std::pair<diag_id, std::vector<std::string>>> original_diag_ids_args )
```
[the remark this answers is not in the document]
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp]: Done.

File: lib/lorina/lorina/diagnostics.hpp [hunk 3]
```diff
@@ -109,11 +119,19 @@ class diagnostic_engine
    */
   void emit_custom_diagnostic( diag_id id, std::vector<std::string> const& args ) const;
 
+  /*! \brief Collect all ids and messages
+   *
+   * \param original_diag_ids_args All pairs of the (id, message)
+   */
+  void collect_messages( const std::vector<std::pair<diag_id, std::vector<std::string>>>& original_diag_ids_args );
```
Comment (marcelwa) [inline lib/lorina/lorina/diagnostics.hpp]:
```suggestion
  void collect_messages( std::vector<std::pair<diag_id, std::vector<std::string>>> const& original_diag_ids_args );
```
To be consistent with `mockturtle`'s east-`const`

File: lib/lorina/lorina/diagnostics.hpp [hunk 4]
```diff
@@ -109,11 +119,19 @@ class diagnostic_engine
...
+
 protected:
   diagnostic_consumer *client_ = nullptr;  /*!< Diagnostic client. */
   std::vector<desc_type> custom_diag_info; /*!< Custom diagnostics. */
   std::map<desc_type, diag_id> custom_diag_ids; /*!< Map from custom ID to diagnostic. */
   mutable uint64_t num_diagnostics{0};
+  std::vector<std::pair<diag_id, std::vector<std::string>>> original_diag_ids_args;
+  std::string all_messages;
```
Comment (marcelwa) [inline lib/lorina/lorina/diagnostics.hpp]: Could you add brief comments like is done above?

File: lib/lorina/lorina/diagnostics.hpp [hunk 5]
```diff
@@ -320,4 +346,28 @@ inline void diagnostic_engine::emit( diag_id id, std::vector<std::string> const&
...
 
+inline void diagnostic_engine::collect_single_msg( diag_id id, std::vector<std::string> const& args )
+{
+    std::pair<diag_id, std::vector<std::string>> collects_ids_args;
+    collects_ids_args.first = id;
+    collects_ids_args.second = args;
+    original_diag_ids_args.push_back( collects_ids_args );
+}
+
+inline void diagnostic_engine::collect_messages( const std::vector<std::pair<diag_id, std::vector<std::string>>>& original_diag_ids_args )
```
Comment (marcelwa) [inline lib/lorina/lorina/diagnostics.hpp]:
```suggestion
inline void diagnostic_engine::collect_messages( std::vector<std::pair<diag_id, std::vector<std::string>>> const& original_diag_ids_args )
```
As above

File: lib/lorina/lorina/diagnostics.hpp [hunk 6]
```diff
@@ -320,4 +346,28 @@ inline void diagnostic_engine::emit( diag_id id, std::vector<std::string> const&
...
+{
+  for ( auto i = 0; i < original_diag_ids_args.size(); i++ )
+  {
+    auto item = original_diag_ids_args[i];
```
Comment (marcelwa) [inline lib/lorina/lorina/diagnostics.hpp]: `item` can be `const`

File: lib/lorina/lorina/diagnostics.hpp [hunk 7]
```diff
@@ -320,4 +346,28 @@ inline void diagnostic_engine::emit( diag_id id, std::vector<std::string> const&
...
+    lorina::diagnostic_level const level = lorina::diag_info[uint32_t( item.first )].first;
+    std::string message = lorina::diag_info[uint32_t( item.first )].second;
+    auto args = item.second;
```
Comment (marcelwa) [inline lib/lorina/lorina/diagnostics.hpp]: `args` can be `const`

File: lib/lorina/lorina/diagnostics.hpp
Comment (marcelwa) [inline lib/lorina/lorina/diagnostics.hpp:1]: There are currently no tests for the new functionality. Since you added it to `lorina`, the obvious solution would be to test it there. However, due to how `mockturtle` handles library includes (copy-paste), backporting is cumbersome. (At some point, all external modules should be included as git submodules.)

Alternatively, could you maybe add tests to the `mockturtle` parsers that test for the newly added functionality?

**Review by marcelwa — CHANGES_REQUESTED** [review]

**Comment by marcelwa** [comment]
@wjrforcyber what is the status on this PR? Are you still working on it?

**Comment by wjrforcyber** [comment]
Sorry I totally forgot about this, I do remember that I had some concerns but I couldn't clearly list them now. I'll pick up this PR and reply to this by the end of this week.

File: lib/lorina/lorina/diagnostics.hpp (hunk 3 above)
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp]: Done.

File: lib/lorina/lorina/diagnostics.hpp (hunk 4 above)
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp]: Done.

File: lib/lorina/lorina/diagnostics.hpp (hunk 5 above)
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp]: Done.

File: lib/lorina/lorina/diagnostics.hpp (hunk 6 above)
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp]: Done.

File: lib/lorina/lorina/diagnostics.hpp (hunk 7 above)
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp]: Done.

File: lib/lorina/lorina/diagnostics.hpp
Comment (wjrforcyber) [reply lib/lorina/lorina/diagnostics.hpp:1]: Done.
However, there are some issues, ~~such as `latch` does not seem to be supported here~~. And there might be too many corner cases. I have added some tests for header, input, output, I think if we would like this to be fully compact we need to list all the corner condition we could come up with.
~~And for latch, it's not supported and it return the error inside the assert~~
https://github.com/marcelwa/mockturtle/blob/31f338a91fb7776748c31b47026747a54b10e96e/include/mockturtle/io/aiger_reader.hpp#L136-L139
~~which I don't like, such error might probably be tricky to redirect to the diagnostic engine.~~

**Comment by marcelwa** [comment]
@wjrforcyber I'm sorry for leaving this PR stale for so long. Are you still interested in pushing it over the finish line? No worries if not. But Copilot seems to still have some remarks.

**Comment by wjrforcyber** [comment]
> @wjrforcyber I'm sorry for leaving this PR stale for so long. Are you still interested in pushing it over the finish line? No worries if not. But Copilot seems to still have some remarks.

Hi! 
Mockturtle is no longer actively maintained and things need to be modified on both sides. Maybe just make it stale or close it.

**Comment by marcelwa** [comment]
Since the EPFL repo is no longer maintained, I use my personal fork in any project that requires it. I'm okay with making whatever changes to it that facilitate my use cases. You're warmly invited to make any edits you see fit.
